mountpoint-s3/src/main.rs

```diff
@@ -12,92 +10,147 @@ use fuser::{MountOption, Session};
+use mountpoint_s3::metrics::MetricsSink;
 use mountpoint_s3::prefix::Prefix;
 use mountpoint_s3_client::{
     AddressingStyle, Endpoint, HeadBucketError, ObjectClientError, S3ClientConfig, S3CrtClient,
 };
 use mountpoint_s3_client::{ImdsCrtClient, S3ClientAuthConfig};
-use mountpoint_s3_crt::common::rust_log_adapter::RustLogAdapter;
 use nix::sys::signal::Signal;
 use nix::unistd::ForkResult;
 use regex::Regex;
-use time::format_description::FormatItem;
-use time::macros;
-use time::OffsetDateTime;
-use tracing_subscriber::{
-    filter::EnvFilter, filter::LevelFilter, layer::SubscriberExt, util::SubscriberInitExt, Layer,
-};
 
 mod build_info;
 
-fn init_tracing_subscriber(is_foreground: bool, log_directory: Option<&Path>) -> anyhow::Result<()> {
-    const LOG_DIRECTORY: &str = ".mountpoint-s3";
-    const LOG_FILE_NAME_FORMAT: &[FormatItem<'static>] =
-        macros::format_description!("mountpoint_s3_[year][month][day][hour][minute][second].log");
+mod logging {
+    use super::*;
+
+    use std::backtrace::Backtrace;
+    use std::fs::{DirBuilder, OpenOptions};
+    use std::os::unix::fs::DirBuilderExt;
+    use std::panic::{self, PanicInfo};
+    use std::thread;
+
+    use mountpoint_s3::metrics::metrics_tracing_span_layer;
+    use mountpoint_s3_crt::common::rust_log_adapter::RustLogAdapter;
+    use time::format_description::FormatItem;
+    use time::macros;
+    use time::OffsetDateTime;
+    use tracing_subscriber::{
+        filter::EnvFilter, filter::LevelFilter, layer::SubscriberExt, util::SubscriberInitExt, Layer,
+    };
 
-    /// Create the logging config from the RUST_LOG environment variable or the default config if
-    /// that variable is unset. We do this in a function because [EnvFilter] isn't [Clone] and we
-    /// need a second copy of it in the foreground case to replicate logs to stdout.
-    fn create_env_filter() -> EnvFilter {
-        EnvFilter::try_from_default_env().unwrap_or_else(|_| EnvFilter::new("info,awscrt=off,fuser=error"))
+    /// Set up all our logging infrastructure.
+    ///
+    /// This method:
+    /// - initializes the `tracing` subscriber for capturing log output
+    /// - sets up the logging adapters for the CRT and for metrics
+    /// - installs a panic hook to capture panics and log them with `tracing`
+    pub(super) fn init_logging(is_foreground: bool, log_directory: Option<&Path>) -> anyhow::Result<()> {
+        init_tracing_subscriber(is_foreground, log_directory)?;
+        install_panic_hook();
+        Ok(())
     }
-    let env_filter = create_env_filter();
 
-    // Don't create the files or subscribers if we'll never emit any logs
-    if env_filter.max_level_hint() == Some(LevelFilter::OFF) {
-        return Ok(());
+    fn tracing_panic_hook(panic_info: &PanicInfo) {
+        let location = panic_info
+            .location()
+            .map(|l| format!("{}", l))
+            .unwrap_or_else(|| String::from("<unknown>"));
+
+        let payload = panic_info.payload();
```

Comment:
What is a `payload()` for the panic info? Could not understand it exactly from documentation. Is it like if panic string is provided it will act as payload. But that is `panic_info` itself?

Reply:
Yeah, payload is the string you passed into the `panic!`, after formatting it (so the payload for `panic!("the number is {}", 42")` is a `String` with the value "the number is 42").